### ISSUE TITLE
Bug: TypeScript version check gets stuck on 5.1.3 even after downgrade

### ISSUE THREAD

**Issue by Mitsunee** [typescript-eslint/typescript-eslint#7102]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Repro Code

Cloned [Foxkit library template](https://github.com/foxkit-js/library-template), and ran the following after replacing `eslint-config-foxkit@next` with a local build that I wish to test, then ran:

```sh
pnpm lint # alias for eslint .
pnpm remove typescript # removes 5.1.x
pnpm add -D typescript@5.0.4 # installs fixed version
pnpm lint
```

### ESLint Config

cannot provide a single config as I'm testing functionality of an entire package of configs that is in development

### tsconfig

Copied from [Foxkit library template](https://github.com/foxkit-js/library-template/blob/main/tsconfig.json)

### Expected Result

No warning about version as I am using 5.0.4, which is supported

### Actual Result

Warning telling me I'm using a version that I am not using

### Additional Info

![Screenshot_20230614_123830](https://github.com/typescript-eslint/typescript-eslint/assets/19474909/d5267753-da24-42ea-9f20-886c38f88e72)

I have tried to delete both node_modules and lockfile and reinstalled, but the issue persists. I have confirmed in multiple ways that I am indeed running TS version 5.0.4. Only deleting the entire project fixed the issue temporarily (read: for about 5mins). I also tried upgrading to `5.59.11` which did not address the issue.

Edit: formatting + more details

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.59.8` and `5.59.11` |
| `@typescript-eslint/parser`            | `5.59.8` and `5.59.11` |
| `@typescript-eslint/scope-manager`     | `5.59.8` and `5.59.11` |
| `@typescript-eslint/typescript-estree` | `5.59.8` and `5.59.11` |
| `@typescript-eslint/type-utils`        | `5.59.8` and `5.59.11` |
| `@typescript-eslint/utils`             | `5.59.8` and `5.59.11` |
| `TypeScript`                           | accidentally ran with `5.1.3` once, `5.0.4` since then |
| `ESLint`                               | `8.41.0` |
| `node`                                 | `18.16.0` |

**Comment by bradzacher** [comment]
If you're getting a warning... then that means that TS5.1 is still installed!
Our tooling has a peer dependency on TS only, and all we do is `require('typescript')` - so if you're seeing a warning logged - then TS5.1 is still installed in some way.
This is likely to do with your package manager and lock files, I'd assume.

https://typescript-eslint.io/linting/troubleshooting/#how-do-i-check-to-see-what-versions-are-installed

**Comment by Mitsunee** [comment]
![Screenshot_20230614_170347](https://github.com/typescript-eslint/typescript-eslint/assets/19474909/dc8e8ae7-b0a4-4b81-a71d-0f4f252fb9b7)

like I said, I tried every possible way I know to confirm I have 5.0.4 installed in the project, yet I keep getting the warning saying I have 5.1.3 installed. Even `require("typescript").version returns `'5.0.4'`
